aocdbc/day_4_passport_processing.py

```diff
@@ -0,0 +1,107 @@
+from re import fullmatch
+import sys
+from typing import Dict, List, Optional
+
+from icontract import require, ensure
+
+_REQUIRED_KEYS = set(["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"])
+
+
+@ensure(lambda: True)
+def blank_line_split(text: str) -> List[str]:
+    # NOTE: I tried to do a lookahead for the 2nd newline, but split doesn't take
+    # a regex.
+    return text.split("\n\n")
+
+
+@ensure(lambda: True)
+def parse(passport: str) -> Optional[Dict[str, str]]:
+    result = {}
+    for tok in passport.split():
+        pair = tok.split(":", 2)
+        if len(pair) != 2:
+            return None
+        result[pair[0]] = pair[1]
+    return result
+
+
+@ensure(lambda entry, result: result == (not (_REQUIRED_KEYS - entry.keys())))
+def is_valid(entry: Dict[str, str]) -> bool:
+    return len(_REQUIRED_KEYS - entry.keys()) == 0
+
+
+@ensure(lambda result: result >= 0)
+def count_valid(batch: str) -> int:
```

Comment:
I think there might be a potential bug. You would expect no new lines in the `passport_text`, but that could still happen with triple new lines. It's probably easier if you have `blank_line_split` operate on `List[str]` as lines (with no new lines as requirement!).

You can introduce, for example:
```
class Lines(DBC, List[str]):
    @require(lambda lines: all('\n' not in line for line in lines))
    def __init__(self, lines: List[str]) -> None:
        list.__init__(self, lines)
```

Reply:
As it happens, `text.split()` breaks on any amount of whitespace, including newlines; e.g.
```
>>> ' \n abc def'.split()
['abc', 'def']
```
So I think what's there is ok. For what it's worth, there is clearly an underspecification in the problem about how we are to interpret unexpected newline counts! In the current implementation, any number of newlines should be counted as a single break.

Reply:
Let's discuss this as part of #25 (maybe also in a call as it might be a bit faster to resolve). It would be cool to come up with a general design pattern for all the exercises to make it easier for the reader to follow.